[#1078] Provide a clean dark theme for the editor

echebbi opened the pull request:
## Contribution

Closes eclipse/xtext#2263 by providing default color preferences when Eclipse IDE uses the dark theme.

These color preferences are similar to JDT's ones (see examples below). I tested on both Eclipse 2020-06 (latest release at the time of writing) and Eclipse Photon.

## Implementation

 - the new `org.eclipse.xtend.ide.themes` bundle provides the colors to use in dark theme as CSS properties and contributes to the _org.eclipse.e4.ui.css.swt.theme_ extension point to make these colors available to Eclipse IDE
 - the symbolic name of the `org.eclipse.xtend.ide.themes` bundle is actually the name of the Xtext grammar (org.eclipse.xtend.core.Xtend) because otherwise Eclipse's CSS engine doesn't override the color preferences when switching to dark theme in Photon

To the best of my knowledge that's the only way to handle a dark theme in Xtext-based editors, see [this thread](https://www.eclipse.org/forums/index.php?t=msg&th=1101764&goto=1821997&#msg_1821997) for details.

I suppose that this new bundle should be added to the Maven/Gradle build and added to a feature but I am unsure about what to do exactly.

## Examples
<div align="center">

_Code snippet:_

![image](https://user-images.githubusercontent.com/25926433/86516904-d7fcf200-be24-11ea-94b4-0acff0720816.png)

_Switching to dark theme:_

![demo-xtend-dark-theme](https://user-images.githubusercontent.com/25926433/86516918-f95dde00-be24-11ea-82f6-f8c8ee9da817.gif)

</div>

echebbi committed:
[#1078] Provide a clean dark theme for the editor

 - the symbolic name of the new 'org.eclipse.xtend.ide.themes' bundle is
   the name of the Xtext grammar ('org.eclipse.xtend.core.Xtend') because
   otherwise Eclipse's CSS engine doesn't override color preferences when
   switching to dark theme in Photon

Signed-off-by: Emmanuel Chebbi <emmanuel.chebbi@outlook.fr>

cdietrich commented:
TODO

- add pom
- add to project parent pom
- add to feature/update site in this repo
- add to update site + feature in xtext-umbrella

- test with oxygen
- check if publishing/deploy needs update

- also: what check else could be missing

echebbi committed:
add about.html

Signed-off-by: Emmanuel Chebbi <emmanuel.chebbi@outlook.fr>

echebbi committed:
Add to Maven build + p2

Signed-off-by: Emmanuel Chebbi <emmanuel.chebbi@outlook.fr>

echebbi commented:
- added about.html
 - added a POM (not sure about the content though: just copied/pasted it from the .ide bundle)
 - added references in parent POM & category.xml (the p2 embeds the new .ide.themes bundle but shows "pluginName" instead of "Xtend IDE Themes" — I haven't investigated why yet)
 - tested with Oxygen

> feature in xtext-umbrella

I have to create a PR in the [xtext-umbrella](https://github.com/eclipse/xtext-umbrella) repository to add the new plug-in to the [releng/org.eclipse.xtend.sdk.feature/feature.xml](https://github.com/eclipse/xtext-umbrella/blob/a91bcbe6d9c546d770aa95d6f135d090e1b1f444/releng/org.eclipse.xtend.sdk.feature/feature.xml#L29-L39), right?

> check if publishing/deploy needs update

Do I have anything to do here?

> cant/should this simply be packaged into the normal .ide bundle

I don't think that's possible. As [explained here](https://www.eclipse.org/forums/index.php?t=msg&th=1101764&goto=1821997&#msg_1821997) the symbolic name of the bundle must match the preferences node used for the Xtext grammar.

kthoms commented:
> I have to create a PR in the xtext-umbrella repository to add the new plug-in to the releng/org.eclipse.xtend.sdk.feature/feature.xml feature, right?

Yes, when it is really necessary to have a separate bundle. I'll check your links.

> check if publishing/deploy needs update
> 
> Do I have anything to do here?

No, this is my task. I'll test & review your PR soon. Thanks so far!

kthoms commented on org.eclipse.xtend.ide.themes/.classpath:
```diff
@@ -0,0 +1,7 @@
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src"/>
```
This project does not contain any Java source. Remove all Java dependencies.
<img width="996" alt="Bildschirmfoto 2020-07-06 um 18 14 53" src="https://user-images.githubusercontent.com/265597/86615574-e843d700-bfb4-11ea-8863-d60cfa6d3ab7.png">

kthoms commented on org.eclipse.xtend.ide.themes/.project:
```diff
@@ -0,0 +1,28 @@
+	</projects>
+	<buildSpec>
+		<buildCommand>
+			<name>org.eclipse.jdt.core.javabuilder</name>
```
remove

kthoms commented on org.eclipse.xtend.ide.themes/.project:
```diff
@@ -0,0 +1,28 @@
+	</buildSpec>
+	<natures>
+		<nature>org.eclipse.pde.PluginNature</nature>
+		<nature>org.eclipse.jdt.core.javanature</nature>
```
remove

echebbi replied:
I kept the PluginNature because the plug-in contributes an extension. Is it unnecessary though?

kthoms commented on org.eclipse.xtend.ide.themes/.settings/org.eclipse.jdt.core.prefs:
```diff
@@ -0,0 +1,8 @@
+eclipse.preferences.version=1
```
remove file

kthoms commented on org.eclipse.xtend.ide.themes/build.properties:
```diff
@@ -0,0 +1,6 @@
+source.. = src/
```
remove

kthoms commented on org.eclipse.xtend.ide.themes/plugin.properties:
```diff
@@ -0,0 +1,12 @@
+###############################################################################
+# Copyright (c) 2020 itemis AG (http://www.itemis.eu) and others.
```
please use your copyright identifier

kthoms commented on org.eclipse.xtend.ide.themes/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,9 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: %pluginName
+Bundle-SymbolicName: org.eclipse.xtend.core.Xtend;singleton:=true
```
change to org.eclipse.xtend.ide.themes
must match the directory name

echebbi replied:
Xtend must have a bundle which symbolic name matches the name of the Xtext grammar otherwise the color preferences won't be taken into account in Oxygen and Photon. The only alternative I see would be to rename the directory instead (so "org.eclipse.xtend.core.Xtend" instead of "org.eclipse.xtend.ide.themes"). Would that suit you?

kthoms replied:
Leave it for now. I‘ll check that requirement. I believe that you investigated in deep, just want to make sure that there is no other way.

kthoms replied:
Leave it for now. I‘ll check that requirement. I believe that you investigated in deep, just want to make sure that there is no other way.

kthoms commented on org.eclipse.xtend.ide.themes/pom.xml:
```diff
@@ -0,0 +1,64 @@
+		<relativePath>../releng/org.eclipse.xtend.tycho.parent</relativePath>
+	</parent>
+	<packaging>eclipse-plugin</packaging>
+	<artifactId>org.eclipse.xtend.core.Xtend</artifactId>
```
change to org.eclipse.xtend.ide.themes

kthoms commented on releng/org.eclipse.xtend.p2-repository/category.xml:
```diff
@@ -12,5 +12,6 @@
    <bundle id="org.eclipse.xtend.m2e.source" version="2.23.0.qualifier"><category name="Xtext"/></bundle>
    <bundle id="org.eclipse.xtend.standalone" version="2.23.0.qualifier"><category name="Xtext"/></bundle>
    <bundle id="org.eclipse.xtend.standalone.source" version="2.23.0.qualifier"><category name="Xtext"/></bundle>
+   <bundle id="org.eclipse.xtend.core.Xtend" version="2.23.0.qualifier"><category name="Xtext"/></bundle>
```
change to org.eclipse.xtend.ide.themes

kthoms reviewed (changes requested):
I had problems starting my Oxygen dev env ATM, so could not do the functional testing yet. Will do this ASAP.
These are formal change requests for now.

echebbi committed:
Change copyright

Signed-off-by: Emmanuel Chebbi <emmanuel.chebbi@outlook.fr>

echebbi committed:
Remove unnecessary dependencies to JDT

Signed-off-by: Emmanuel Chebbi <emmanuel.chebbi@outlook.fr>

kthoms commented:
TP: 2020-06

When switching back to light theme the colors stay. If we follow this path I guess a light theme is needed, too.
<img width="526" alt="Bildschirmfoto 2020-07-06 um 21 32 07" src="https://user-images.githubusercontent.com/265597/86632960-58138b00-bfd0-11ea-8e99-aaf8b4a5bf11.png">

echebbi commented:
> When switching back to light theme the colors stay

Yes, but not all and others are "corrupted" (e.g. multi-line comments become blue when switching to dark theme). I've always witnessed such behavior on Xtext-based editors; restarting the editor/IDE does the trick.

> If we follow this path I guess a light theme is needed, too.

It's worth a try. It seems that JDT [doesn't provide one](https://git.eclipse.org/c/jdt/eclipse.jdt.ui.git/tree/org.eclipse.jdt.ui/css) so there may be an alternative. I'll update the PR with a light theme inspired from [XtendHighlightingConfiguration](https://github.com/eclipse/xtext-xtend/blob/5799154cd8fced549cdc0eca56e57d23f3feb523/org.eclipse.xtend.ide/src/org/eclipse/xtend/ide/highlighting/XtendHighlightingConfiguration.java#L21) and [XbaseHighlightingConfiguration](https://github.com/eclipse/xtext-eclipse/blob/d19246cb2fbc057aa8793a53795301838ab2a1d0/org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/highlighting/XbaseHighlightingConfiguration.java#L24).

kthoms commented:
Please try the changes on this branch: https://github.com/eclipse/xtext-xtend/tree/echebbi-1078-provide-a-clean-dark-theme
I have integrated your CSS into Xtend's dark style CSS and applied the overridden preferences on theme change. I really would like to avoid the introduction of an additional bundle.
This looks like what I've seen with your additional bundle. I did not test on Oxygen TP yet.
The rich strings don't look properly colored yet. 
<img width="1314" alt="Bildschirmfoto 2020-07-07 um 06 33 54" src="https://user-images.githubusercontent.com/265597/86712170-03e2c800-c01d-11ea-8a2c-f5c29de668bc.png">

kthoms commented:
@echebbi Would be great if you find time to check my approach

kthoms commented:
Hey @echebbi. It would be really great to get your feature into the next release. Do you think you can work on the mentioned issues (switch to light theme, combine into xtend.ide bundle)?

echebbi commented:
Sorry @kthoms: I have been very busy lately. I'll check your changes as soon as possible.

kthoms commented:
No problem, good to hear you want to finish this. I'm on vacation from now on, could also be that you get a delayed answer then.

cdietrich commented:
move to 2.24

cdietrich commented:
replaced by https://github.com/eclipse/xtext-xtend/pull/1138
